packages/insomnia-app/app/ui/components/editors/environment-editor.js

```diff
@@ -10,18 +10,40 @@ import { NUNJUCKS_TEMPLATE_GLOBAL_PROPERTY_NAME } from '../../../templating';
 // Docs: https://github.com/DeNA/nedb#inserting-documents
 const INVALID_NEDB_KEY_REGEX = /^\$|\./;
 
-export const ensureKeyIsValid = (key: string): string | null => {
+export const ensureKeyIsValid = (key: string, isRoot: boolean = false): string | null => {
   if (key.match(INVALID_NEDB_KEY_REGEX)) {
     return `"${key}" cannot begin with '$' or contain a '.'`;
   }
 
-  if (key === NUNJUCKS_TEMPLATE_GLOBAL_PROPERTY_NAME) {
+  if (key === NUNJUCKS_TEMPLATE_GLOBAL_PROPERTY_NAME && isRoot) {
     return `"${NUNJUCKS_TEMPLATE_GLOBAL_PROPERTY_NAME}" is a reserved key`; // verbiage WIP
   }
 
   return null;
 };
 
+/**
+ * Recursively check keys of nested object and immediately return when found
+ * @param {object} obj - object to analyse
+ * @returns {boolean} - if any invalid key is found
+ */
+export function checkNestedKeys(obj: any, isRoot: boolean = true): string | null {
+  let result = null;
+  for (const key in obj) {
+    // Nested keys
+    if (typeof obj[key] === 'object' && obj.hasOwnProperty(key)) {
+      result = checkNestedKeys(obj[key], false);
+    } else {
+      result = ensureKeyIsValid(key, isRoot);
+    }
+    if (result !== null) {
+      break;
+    }
+  }
+
+  return result;
```

Comment:
It should be safe to simply exit without a null check, since `checkNestedKeys` has a return type that includes `null`. In addition, further down, there is a check to only set an error if an error is returned, so we don't need to double-up checks.

Also, it feels as though the `obj.hasOwnProperty(key)` check is obsolete, because `typeof obj[key]` returns `undefined` if the key is not present. The second condition will never be reached.

```suggestion
  for (const key in obj) {
    // Nested keys
    if (typeof obj[key] === 'object') {
      return checkNestedKeys(obj[key], false);
    } else {
      return ensureKeyIsValid(key, isRoot);
    }
  }
```